Nixpkgs documentation: reword info about version testers

AndersonTorres opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

AndersonTorres committed:
doc/stdenv/passthru.chapter.md: expand Package tests section

- Reword the paragraphs
- Remove the comparison with `versionCheckHook`
  - It will be moved to  `versionCheckHook.section.md`

AndersonTorres committed:
doc/build-helpers/testers.chapter.md: expand testVersion

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4695

Daholli reviewed (approved):
Sounds very good, I cannot really judge the content besides being able to understand it better

doronbehar reviewed (commented):
Thanks for the PR. The rephrasings are indeed better, but I'm not sure about a few things.

AndersonTorres commented on doc/stdenv/passthru.chapter.md:
```diff
@@ -75,17 +75,16 @@ The Nixpkgs systems for continuous integration [Hydra](https://hydra.nixos.org/)
-- If the `versionCheckPhase` (the phase defined by [`versionCheckHook`](#versioncheckhook)) fails, it triggers a failure which can't be ignored if you use the package, or if you find out about it in a [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review) report.
-- Sometimes packages become silently broken - meaning they fail to launch but their build passes because they don't perform any tests in the `checkPhase`. If you use this tool infrequently, such a silent breakage may rot in your system / profile configuration, and you will not notice the failure until you will want to use this package. Testing such basic functionality ensures you have to deal wit ...
-- When you open a PR, [ofborg](https://github.com/NixOS/ofborg)'s CI _will_ run `passthru.tests` of [packages that are directly changed by your PR (according to your commits' messages)](https://github.com/NixOS/ofborg?tab=readme-ov-file#automatic-building), but if you'd want to use the [`@ofborg build`](https://github.com/NixOS/ofborg?tab=readme-ov-file#build) command for dependent packages, you  ...
+It is also possible to use `passthru.tests` to test the version with [`testers.testVersion`](#tester-testVersion).
```
I reordered the bullets to make it more readable.

AndersonTorres commented on doc/stdenv/passthru.chapter.md:
```diff
@@ -75,17 +75,16 @@ The Nixpkgs systems for continuous integration [Hydra](https://hydra.nixos.org/)
 
-It is also possible to use `passthru.tests` to test the version with [`testVersion`](#tester-testVersion), but since that is pretty trivial and recommended thing to do, we recommend using [`versionCheckHook`](#versioncheckhook) for that, which has the following advantages over `passthru.tests`:
-
-- If the `versionCheckPhase` (the phase defined by [`versionCheckHook`](#versioncheckhook)) fails, it triggers a failure which can't be ignored if you use the package, or if you find out about it in a [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review) report.
-- Sometimes packages become silently broken - meaning they fail to launch but their build passes because they don't perform any tests in the `checkPhase`. If you use this tool infrequently, such a silent breakage may rot in your system / profile configuration, and you will not notice the failure until you will want to use this package. Testing such basic functionality ensures you have to deal wit ...
-- When you open a PR, [ofborg](https://github.com/NixOS/ofborg)'s CI _will_ run `passthru.tests` of [packages that are directly changed by your PR (according to your commits' messages)](https://github.com/NixOS/ofborg?tab=readme-ov-file#automatic-building), but if you'd want to use the [`@ofborg build`](https://github.com/NixOS/ofborg?tab=readme-ov-file#build) command for dependent packages, you  ...
```
Comparing both methods on a third place is a nice idea.
I will do this asap.

doronbehar replied:
> > Hmm in principal I don't mind much how this got moved and rephrased into `versionCheckHook.section.md`, but I think there should be a single place in which we compare both of the methods. Perhaps the "Package tests" subtitle in `pkgs/README.md` is a better? What do you think? If you agree we should concentrate the information about this somewhere, I hope we'd agree the correct thing to do would be to link from here and from `versionCheckHook.section.md` to that comparison section.
>
> Comparing both methods on a third place is a nice idea. I will do this asap.

One more thing: I don't like how this section was moved to a different place on a different commit (I'm reviewing this PR commit-wise).

AndersonTorres replied:
Modifying multiple files at once is confusing and is prone to merge conflicts.

AndersonTorres commented on doc/build-helpers/testers.chapter.md:
```diff
@@ -169,13 +169,14 @@ The build will fail if `shellcheck` finds any issues.
 Checks that the output from running a command contains the specified version string in it as a whole word.
 
-NOTE: In most cases, [`versionCheckHook`](#versioncheckhook) should be preferred, but this function is provided and documented here anyway. The motivation for adding either tests would be:
+Although being very simple and naive, this test is useful:
```
If the idea is to refer to the hook, the end of the whole paragraph is better, so it does not break the reading flow.

AndersonTorres replied:
> Is that intended?

Yes. Both are useful in their own sight.

> Should both be implemented if possible?

There are no strong requirements for implementing any sort of tests. 
It is up to any Nixpkgs contributor to decide for their inclusion.

AndersonTorres commented on pkgs/README.md:
```diff
@@ -739,6 +739,32 @@ stdenv.mkDerivation {
+
+- Since [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review) does not execute `passthru.tests` attributes, it will not execute `passthru.tests.version`.
+
+  On the other hand, `versionCheckPhase` will be executed as a regular phase during the build time of the derivation, therefore it will be indirectly catched by `nixpkgs-review`, working around this limitation.
```
Reworded.

Aleksanaa commented on pkgs/README.md:
```diff
@@ -741,6 +741,34 @@ stdenv.mkDerivation {
+
+Many programs provide a command-line flag that emits version information. It provides a straightforward test procedure, namely, run the program with that specific flag and verify the output.
+
+Nixpkgs provides two tools for this task: [`testers.testVersion`](#tester-testVersion) and [`versionCheckHook`](#versioncheckhook).
```
This kind of tags won't work outside of nixpkgs documentation

AndersonTorres replied:
"Fixed".

Aleksanaa commented on pkgs/README.md:
```diff
@@ -741,6 +741,34 @@ stdenv.mkDerivation {
+
+The advantages of `testers.testVersion` over `versionCheckHook` are listed at [Package tests](#var-passthru-tests-packages) already. Below we will list some advantages of `versionCheckHook` over `testers.testVersion`:
+
+- Since [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review) does not execute `passthru.tests` attributes, it will not execute `passthru.tests.version`.
+
+  On the other hand, `versionCheckHook` will be executed as a regular phase during the build time of the derivation, therefore it will be indirectly catched by `nixpkgs-review`, working around this limitation.
+
+- When a pull request is opened against Nixpkgs repository, [ofborg](https://github.com/NixOS/ofborg)'s CI will automatically run `passthru.tests` attributes for the packages that are [directly changed by your PR (according to your commits' messages)](https://github.com/NixOS/ofborg?tab=readme-ov-file#automatic-building).
+
+  However, ofBorg does not run the `passthru.tests` attributes for _transitive dependencies_ of those packages. To execute them, commands like [`@ofborg build dependency1.tests dependency2.tests ...`](https://github.com/NixOS/ofborg?tab=readme-ov-file#build) are needed.
+
+  On the other hand, `versionCheckHook` will be executed as a regular phase during the build time of the derivation, therefore it will be indirectly catched by ofBorg, working around this limitation.
+
+- Sometimes a package triggers no errors while being built - especially when the upstream provides no unit tests-, however it fails at runtime. When such a package is not used in a regular basis, it leaves room for a silent breakage that rots in the system / profile configuration, not being noticed for long periods of time until the next usage of this package.
+
+  Although `passthru.tests` fills the same purpose, it is more prone to be forgotten by human beings; on the other hand, `versionCheckHook` will be executed as a regular phase during the build time of the derivation, therefore it will not be accidentally ignored.
+
+Despite having an almost identical functioning and a huge overlap, `versionCheckHook` and `testers.testVersion` have complementary roles, and there are no impediments for using both at the same time.
```
These should be moved back to nixpkgs documentation.

AndersonTorres replied:
Done.

Aleksanaa commented on pkgs/README.md:
```diff
@@ -741,6 +741,34 @@ stdenv.mkDerivation {
+
+- When a pull request is opened against Nixpkgs repository, [ofborg](https://github.com/NixOS/ofborg)'s CI will automatically run `passthru.tests` attributes for the packages that are [directly changed by your PR (according to your commits' messages)](https://github.com/NixOS/ofborg?tab=readme-ov-file#automatic-building).
+
+  However, ofBorg does not run the `passthru.tests` attributes for _transitive dependencies_ of those packages. To execute them, commands like [`@ofborg build dependency1.tests dependency2.tests ...`](https://github.com/NixOS/ofborg?tab=readme-ov-file#build) are needed.
```
Ofborg also doesn't run `versionCheckHook` for those transisive dependencies, unless you add them to passthru?

doronbehar replied:
> Ofborg also doesn't run `versionCheckHook` for those transisive dependencies, unless you add them to passthru?

If `passthru.tests` is empty, indeed ofborg won't run `versionCheckHook` for those dependent packages, but nixpkgs-review will.

Aleksanaa commented on pkgs/README.md:
```diff
@@ -741,6 +741,34 @@ stdenv.mkDerivation {
+
+  However, ofBorg does not run the `passthru.tests` attributes for _transitive dependencies_ of those packages. To execute them, commands like [`@ofborg build dependency1.tests dependency2.tests ...`](https://github.com/NixOS/ofborg?tab=readme-ov-file#build) are needed.
+
+  On the other hand, `versionCheckHook` will be executed as a regular phase during the build time of the derivation, therefore it will be indirectly catched by ofBorg, working around this limitation.
```
It's worth noting that people are supposed to add one package per commit, so ofborg will still build all `passthru.tests`.

AndersonTorres replied:
This is more about future updates. E. g. SDL has many dependencies but an update of SDL does not trigger the already existing transitive dependencies in the current CI setup.

Aleksanaa commented on pkgs/README.md:
```diff
@@ -741,6 +741,34 @@ stdenv.mkDerivation {
+
+  On the other hand, `versionCheckHook` will be executed as a regular phase during the build time of the derivation, therefore it will be indirectly catched by ofBorg, working around this limitation.
+
+- Sometimes a package triggers no errors while being built - especially when the upstream provides no unit tests-, however it fails at runtime. When such a package is not used in a regular basis, it leaves room for a silent breakage that rots in the system / profile configuration, not being noticed for long periods of time until the next usage of this package.
```
You might consider putting this point at the top

AndersonTorres replied:
Done.

doronbehar commented on doc/build-helpers/testers.chapter.md:
```diff
@@ -169,15 +169,18 @@ The build will fail if `shellcheck` finds any issues.
-- Catch dynamic linking errors and such and missing environment variables that should be added by wrapping.
-- Probable protection against accidentally building the wrong version, for example when using an "old" hash in a fixed-output derivation.
+- It assures the main program can run.
+- It catches dynamic linking errors, missing or malformed environment variables and other runtime errors.
```
You know the word "catches" doesn't fit a `passthru.tests.version` check, because if you won't build that `passthru.tests` derivation you won't catch anything. I'd use the word "catches" when describing `versionCheckHook`, but not when describing both.

AndersonTorres replied:
[You approved that wording before](https://github.com/NixOS/nixpkgs/pull/348025#discussion_r1807456082).

doronbehar replied:
> [You approved that wording before](https://github.com/NixOS/nixpkgs/pull/348025#discussion_r1807456082).

Correct, I inspected your phrasing in more details and changed my mind.

> Further, I don't think this difference matters here.
> 
> `passthru.tests.version` catches errors when executed. `versionCheckHook` catches errors when executed.
> 
> The difference is `passthru.tests.version` is executed directly when called e.g. via commandline, while `versionCheckHook` is executed indirectly when the package containing it is built.

That's a big difference IMO. Of course we agree that if you'd explicitly build `passthru.tests.version` you'd catch an issue with the package, but I find it inaccurate to suggest that merely adding a `passthru.tests.version` will make any such issue be catched, because all the reasons you are already aware of from [our discussion on Discourse](https://discourse.nixos.org/t/improve-the-jdk-infrastructure-on-nixpkgs/45678/21).

AndersonTorres replied:
Please suggest a better phrasing, since this word was present on the document already.

doronbehar replied:
> Please suggest a better phrasing, since this word was present on the document already.

Just to be clear, I wouldn't have said anything if these words weren't present in the diff.. But I'd use the words:

> It can bring to awareness dynamic linking errors, missing or malformed environment variables and other runtime errors.

AndersonTorres replied:
Done.

doronbehar commented on doc/build-helpers/testers.chapter.md:
```diff
@@ -169,15 +169,18 @@ The build will fail if `shellcheck` finds any issues.
-it will check `meta.mainProgram` first, and fall back to `pname` or `name`.
 The default argument to the command is `--version`, and the version to be checked will be inferred from the given `package` attribute as well.
 
+Note: see also [`versionCheckHook`](#versioncheckhook), an independent hook with a similar functionality.
```
I'm missing a reference here to a place where the hooks are compared.

doronbehar commented on doc/hooks/versionCheckHook.section.md:
````diff
@@ -22,14 +24,32 @@ stdenv.mkDerivation (finalAttrs: {
 ```
 
-Note that for [`buildPythonPackage`](#buildpythonpackage-function) and [`buildPythonApplication`](#buildpythonapplication-function), `doInstallCheck` is enabled by default.
+Note that some builders, e.g. [`buildPythonPackage`](#buildpythonpackage-function) and [`buildPythonApplication`](#buildpythonapplication-function), enable `doInstallCheck` by default.
````
It's not such a big effort IMO to maintain here a list of all builders that enable `installCheck` phase by default. Hence, the previous phrasing is a bit more decisive and informative. This phrasing sort of suggests: "We don't keep track of which builders enable `installCheck` by default and which do, so you should inspect the source code of each if you don't want to enable it when not needed."

AndersonTorres replied:
Hum, should this paragraph be rewritten each time a new custom builder appears? I don't think so.
This paragraph looks like an example of builder, not a exhaustive listing.

Further, the documentation of these builders should provide this information, and it is the duty of a contributor to read the documentation of that builder at least.
If this info is not found in the docs, a bug report should be opened and the docs team should be pinged.

doronbehar replied:
> Further, the documentation of these builders should provide this information, and it is the duty of a contributor to read the documentation of that builder at least.
> If this info is not found in the docs, a bug report should be opened and the docs team should be pinged.

I agree with that. But as for:

> Hum, should this paragraph be rewritten each time a new custom builder appears? I don't think so.
> This paragraph looks like an example of builder, not a exhaustive listing.

Nowadays, the list of builders that enable `installCheck` is very small, so I don't find it exhaustive to list them. I agree that theoretically if a new builder that enables `installCheck` will be added in the future to Nixpkgs, probably no one will remember or know to add it to this "list", but I doubt that will happen TBH, from my experience with Nixpkgs.

The previous phrasing also didn't suggest that the builders mentioned is the definitive list of builders that enable `installCheck` by default. I just say that your phrasing slightly more implies that maybe many more builders are enabling `installCheck`, and that is not true.

AndersonTorres replied:
Reverted, in terms. I reworded the example in a more direct sentence.

doronbehar commented on doc/hooks/versionCheckHook.section.md:
```diff
@@ -22,14 +24,32 @@ stdenv.mkDerivation (finalAttrs: {
-- `versionCheckProgramArg`: The argument that needs to be passed to `versionCheckProgram`. If undefined the hook tries first `--help` and then `--version`. Examples: `version`, `-V`, `-v`.
-- `preVersionCheck`: A hook to run before the check is done.
-- `postVersionCheck`: A hook to run after the check is done.
+Note: see also [`testers.testVersion`](#tester-testVersion).
```
Once again, it is a bit annoying IMO to read this and to see that there is a completely different alternative and not having any reference to a comparison between the two.

AndersonTorres replied:
I moved the comparison to this file.

doronbehar commented on pkgs/README.md:
```diff
@@ -741,6 +741,34 @@ stdenv.mkDerivation {
+
+The most notable difference between `versionCheckHook` and `testers.testVersion` is that `versionCheckHook` runs at the build time of the package, while `passthru.tests.versions` executes a whole derivation that depends on the package.
+
+The advantages of `testers.testVersion` over `versionCheckHook` are listed at [Package tests](#var-passthru-tests-packages) already. Below we will list some advantages of `versionCheckHook` over `testers.testVersion`:
```
Hmm it felt natural how the advantages of `testVersion` were listed there, but here it doesn't feel natural at all how the advantages of `versionCheckHook` are listed here and not in it's documentation. If this section is a comparison, it should aggregate advantages and disadvantages of both methods in the same place.

AndersonTorres replied:
[You suggested to put the comparison in pkgs/README](https://github.com/NixOS/nixpkgs/pull/348025#discussion_r1807456582) before.

Do you think this comparison should be in versionCheckHook.md instead?

doronbehar replied:
> [You suggested to put the comparison in pkgs/README](https://github.com/NixOS/nixpkgs/pull/348025#discussion_r1807456582) before.

I know, but I didn't anticipate that it'd come out so inconsistent, because you'd keep the pros and cons of `testVersion` elsewhere. What I suggested was (this time more accurately):

Put the 2x2 matrix of pros & cons of both `testVersion` & `versionCheckHook` in a neutral place like `pkgs/README.md`, and link to that comparison section from both `testVersion`'s documentation and `versionCheckHook.section.md`. Mention shortly the motivation to use either tests in `testVersion`'s documentation and `versionCheckHook.section.md`.

I cannot promise I'd approve such a change never the less, because if we can (and I think we should) reach a consensus that one of the approaches should be the default approach to use, there is no justification for putting such an exhaustive 2x2 matrix of pros and cons in a neutral place.

AndersonTorres replied:
`pkgs/readme.md` does not work.
For now I will move this to versioncheckhook docs

doronbehar commented on pkgs/README.md:
```diff
@@ -741,6 +741,34 @@ stdenv.mkDerivation {
+
+  Although `passthru.tests` fills the same purpose, it is more prone to be forgotten by human beings; on the other hand, `versionCheckHook` will be executed as a regular phase during the build time of the derivation, therefore it will not be accidentally ignored.
+
+Despite having an almost identical functioning and a huge overlap, `versionCheckHook` and `testers.testVersion` have complementary roles, and there are no impediments for using both at the same time.
```
I don't agree with this sentence. Why would someone make an effort to use both? Why can't we agree that either of them fits most cases?

AndersonTorres replied:
> Why would someone make an effort to use both? 

I don't know. Maybe stress-testing.

> Why can't we agree that either of them fits most cases?

Because there is no objective metric for this right now.

On the other hand, I can delete this last line.

doronbehar replied:
> > Why can't we agree that either of them fits most cases?
> 
> Because there is no objective metric for this right now.

Objective metric? Metric of what?

AndersonTorres replied:
Metric of fittingness.

"For packages with such and such characteristics, this approach is well-suited because of such and such premises; otherwise the other approach is better".

doronbehar reviewed (changes requested):
I have a general comment: When such a big system such as Nixpkgs as managed and designed, you want to keep the amount of moving parts to a minimum, especially since you want to attract as much contributors as possible, and to make it easy for them as possible to contribute.

Hence, having 2 perfectly legitimate ways of doing sort of the same thing is a binary moving part that is not justified in my opinion - inexperienced contributors reading these docs won't know what to do and I anticipate their decision will be random in the best case, and they'd stay confused and won't contribute in the worst case. I really think we should reach a decision regarding this.

AndersonTorres committed:
doc/hooks/versionCheckHook.section.md: compare `testers.testVersion`

AndersonTorres committed:
doc/hooks/versionCheckHook.section.md: expand and reword

- Flip the explanation and buildPythonPackage reference paragraphs
- Add a section to emphasize the attributes controlling the hook

doronbehar commented on doc/build-helpers/testers.chapter.md:
```diff
@@ -169,15 +169,21 @@ The build will fail if `shellcheck` finds any issues.
-it will check `meta.mainProgram` first, and fall back to `pname` or `name`.
 The default argument to the command is `--version`, and the version to be checked will be inferred from the given `package` attribute as well.
 
+Note: see also:
+
+- [`versionCheckHook`](#versioncheckhook), an independent hook with similar functionality.
+  - [A comparison between both](#versioncheckhook-comparison-testversion)
```
I find it confusing to mention this alternative and not telling the users right away which approach should be used most of the times.

AndersonTorres replied:
The texts cite each other.
The comparison chapter is the best fit for this info.

doronbehar commented on doc/hooks/versionCheckHook.section.md:
```diff
@@ -22,14 +24,64 @@ stdenv.mkDerivation (finalAttrs: {
 It does so in a clean environment (using `env --ignore-environment`), and it checks for the `${version}` string in both the `stdout` and the `stderr` of the command. It will report to you in the build log the output it received and it will fail the build if it failed to find `${version}`.
 
-The variables that this phase control are:
+Note that `doInstallCheck` is enabled by default for [`buildPythonPackage`](#buildpythonpackage-function) and [`buildPythonApplication`](#buildpythonapplication-function).
```
That's almost the same sentence..

AndersonTorres replied:
Yes, but without a comma "pausing" it.

doronbehar commented on doc/hooks/versionCheckHook.section.md:
```diff
@@ -22,14 +24,64 @@ stdenv.mkDerivation (finalAttrs: {
+  - E.g. extra tools, customized programs etc.
+- `testers.testVersion` executed in an identical environment of a consumer would, independent of the environment it was built.
+- `testers.testVersion` can be executed without rebuilding the package, saving time especially when the rebuild takes too much time.
+- `testers.testVersion` do not add overhead to each build, since it runs after the build, not during it.
```
Calling this an overhead is an overstatement IMO.

AndersonTorres replied:
Objectively the overhead of testVersion is zero.

I will add an extra bullet.

doronbehar reviewed (commented):
Getting better, but I'd still insist that we should tell users that in most cases one of the approaches is better.

doronbehar commented on doc/hooks/versionCheckHook.section.md:
```diff
@@ -22,14 +24,65 @@ stdenv.mkDerivation (finalAttrs: {
-- `versionCheckProgramArg`: The argument that needs to be passed to `versionCheckProgram`. If undefined the hook tries first `--help` and then `--version`. Examples: `version`, `-V`, `-v`.
-- `preVersionCheck`: A hook to run before the check is done.
-- `postVersionCheck`: A hook to run after the check is done.
+Given the current limitations of Nixpkgs's tooling and CI demonstrated above, `versionCheckHook` fits better in most of typical situations.
```
I'm excited to see you wrote this eventually :smiley: .

doronbehar reviewed (commented):
I see only 2 issues left. Looking good.

doronbehar reviewed (changes requested):
OK now I reviewed the PR with the intention of actually perhaps merging it, so I noticed a few more details with which I'm not satisfied. I must say that most of the rephrasing is pretty good, but I feel like this PR tries under the hood to give the `testers.testVersion` and `versionCheckHook` a similar position, and I don't agree with this direction.

Earlier I suggested putting a 2x2 comparison matrix of pros and cons of `testers.testVersion` and `versionCheckHook`. I don't think it is required, because it is more important to give readers the bottom line that (agreeably) states that `versionCheckHook` fits most typical cases, but never the less I'll write here the 2x2 comparison matrix in order to perhaps convince slightly more the author of the PR:

| ****                                                                                                                   | `versionCheckHook`                                                                                                                                                             | `testers.testVersion`                                                                                                                                                                     |
|------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **Special customization**                                                                                              | The hook doesn't provide many ways to control it, other then with attributes document to control it.                                                                           | Can access all features of Nix and Nixpkgs, like generating multiple version test derivations from a list of strings, and accessing programs / tools not part of the derivation's inputs. |
| **Environment**                                                                                                        | Uses `env --ignore-environment --chdir=/` to run the executables in a pure environment as possible, but there is no way to change this behavior.                               | Executed in an identical environment of a consumer, with independently of the environment the executable was built at.                                                                    |
| **Requires rebuild?**                                                                                                  | Yes                                                                                                                                                                            | No                                                                                                                                                                                        |
| **Overhead?**                                                                                                          | Negligible                                                                                                                                                                     | Zero                                                                                                                                                                                      |
| **content-addressed derivations compatible**                                                                           | Can be problematic (TODO: Explain why - I don't understand this part)                                                                                                          | TODO: Should be OK, explain why.                                                                                                                                                          |
| **Being run by @ofborg?**                                                                                              | Yes, as it is part of building a package.                                                                                                                                      | Yes, for directly changed packages, but for dependent packages you'd `@ofborg build`, you'd have to remember to also build the dependent package's `.tests` derivations.                  |
| **Being run by `nixpkgs-review`?**                                                                                     | Yes, as it is part of building a package.                                                                                                                                      | No, but see [this nixpkgs-review PR](https://github.com/Mic92/nixpkgs-review/pull/397).                                                                                                   |
| **What happens when package `A` depending on `B` is all of a sudden builds but does not work due to a change in `B`?** | You will notice a breakage as soon you attempt to build the package - will happen at your next system upgrade etc, and when someone runs a `nixpkgs-review` when changing `B`. | No one will notice the breakage until you will attempt to use the package on your system, because practically no one runs `passthru.tests` regularly.                                     |
| **How likely will you notice a breakage?**                                                                             | Impossible not to notice.                                                                                                                                                      | Very possible, especially if the package is not well maintained and not often used.                                                                                                       |


P.S:

I'd change the commit title of the last commit to:

```
versionCheckHook.section.md: compare with testers.testVersion
```

So it will not span over 2 lines..

AndersonTorres commented on doc/hooks/versionCheckHook.section.md:
````diff
@@ -16,20 +20,73 @@ stdenv.mkDerivation (finalAttrs: {
 })
 ```
 
-Note that for [`buildPythonPackage`](#buildpythonpackage-function) and [`buildPythonApplication`](#buildpythonapplication-function), `doInstallCheck` is enabled by default.
````
This info about an otherwise external actor (namely the Python ecosystem) was inserted among two text blocks that describe the main character of this text, namely the versioncheckhook.

Please look at the solution proposed for this.

AndersonTorres commented on doc/hooks/versionCheckHook.section.md:
```diff
@@ -16,20 +20,73 @@ stdenv.mkDerivation (finalAttrs: {
+- `testers.testVersion` is executed in an identical environment of a consumer would, independent of the environment it was built.
+- `testers.testVersion` can be executed without rebuilding the package, saving time especially when the rebuild takes too long.
+- `testers.testVersion` do not add overhead to each build, since it runs after the build, not during it.
+  - Arguably, the overhead of `versionCheckHook` is negligible, so this point is not much a difference.g
```
Since this was converted into a table, I will not deal with gravestone comments.

AndersonTorres commented:
[And the prize for the best error message goes to](https://github.com/NixOS/nixpkgs/actions/runs/11653896760/job/32446793138?pr=348025)

> 
> 
> Traceback (most recent call last):
>   File "/nix/store/q64nfqryc6kbpk0r34cvcva6c57whzs8-nixos-render-docs-0.0/lib/python3.12/site-packages/nixos_render_docs/__init__.py", line 49, in main
>     manual.run_cli(args)
>   File "/nix/store/q64nfqryc6kbpk0r34cvcva6c57whzs8-nixos-render-docs-0.0/lib/python3.12/site-packages/nixos_render_docs/manual.py", line 704, in run_cli
>     _run_cli_html(args)
>   File "/nix/store/q64nfqryc6kbpk0r34cvcva6c57whzs8-nixos-render-docs-0.0/lib/python3.12/site-packages/nixos_render_docs/manual.py", line 696, in _run_cli_html
>     md.convert(args.infile, args.outfile)
>   File "/nix/store/q64nfqryc6kbpk0r34cvcva6c57whzs8-nixos-render-docs-0.0/lib/python3.12/site-packages/nixos_render_docs/manual.py", line 523, in convert
>     super().convert(infile, outfile)
>   File "/nix/store/q64nfqryc6kbpk0r34cvcva6c57whzs8-nixos-render-docs-0.0/lib/python3.12/site-packages/nixos_render_docs/manual.py", line 41, in convert
>     raise RuntimeError(f"failed to render manual {infile}") from e
> RuntimeError: failed to render manual manual.md
> error:
>         failed to render manual manual.md
> 
> caused by:
>         processing included file build-helpers.md from line 12
> 
> caused by:
>         processing included file hooks/index.md from line 26
> 
> caused by:
>         processing included file hooks/versionCheckHook.section.md from line 38
> 
> caused by:
>

doronbehar commented on doc/build-helpers/testers.chapter.md:
```diff
@@ -169,15 +169,24 @@ The build will fail if `shellcheck` finds any issues.
 Checks that the output from running a command contains the specified version string in it as a whole word.
 
-NOTE: In most cases, [`versionCheckHook`](#versioncheckhook) should be preferred, but this function is provided and documented here anyway. The motivation for adding either tests would be:
+Note: Given the current limitations of Nixpkgs's tooling and CI, `versionCheckHook` fits better than `testers.testVersion` in most of typical situations.
```
Hmm this suggests a sort of dissatisfaction that might be opinionated, but I don't have a strong opinion regarding it, so I won't make this comment of mine a blocker.

AndersonTorres replied:
Indirectly [you recognized the limitations of the tooling](https://github.com/NixOS/nixpkgs/pull/348025#discussion_r1827418529):

> My opinion is that no one should be blamed for not doing something that **our tooling doesn't allow them to easily do**.

doronbehar replied:
> [You recognized the limitations of the tooling](https://github.com/NixOS/nixpkgs/pull/348025#discussion_r1827418529), at least indirectly:
> 
> > My opinion is that no one should be blamed for not doing something that **our tooling doesn't allow them to easily do**.

I agree that these are "limitations", that also don't allow us to easily run `passthru.tests` of packages automatically etc., but I don't necessarily think that these limitations should be solved at some point. I'd say that calling this a limitation might even be an overstatement, at least for some tools. It'd be very nice if `nixpkgs-review` would have built `passthru.tests` automatically, but I wouldn't say that ofborg is not that well designed. Not only that, even if `nixpkgs-review` would have run it, I'd still claim that `versionCheckHook` is better, because running `nixpkgs-review` is not mandatory, and silent breakages can still rot in your system.

doronbehar commented on doc/build-helpers/testers.chapter.md:
```diff
@@ -169,15 +169,24 @@ The build will fail if `shellcheck` finds any issues.
+
+- [`versionCheckHook`](#versioncheckhook), an independent hook with similar functionality.
+  - [A comparison between both](#versioncheckhook-comparison-testversion).
+  - Given the current limitations of Nixpkgs's tooling and CI, `versionCheckHook` fits better than `testers.testVersion` in most of typical situations.
```
You already wrote about the limitations above.

AndersonTorres replied:
You required to put it on different places, usually before a section to save time and after them to summarize conclusions.

doronbehar replied:
> You required to put it on different places, usually before a section to save time and after them to summarize conclusions.

It is correct that I stated that there is no choice but to mention it both `passthru.chapter.md` and `testers.chapter.md`, but I think it should be mentioned in the same file twice, in less then 10 lines distance between each mention.

AndersonTorres replied:
I am failing to understand your requirement here.

You said while commenting this same file `doc/build-helpers/testers.chapter.md`:

> [Would you agree to (after all) write here the bottom line of the comparison? Which is that versionCheckHook fits most typical cases?](https://github.com/NixOS/nixpkgs/pull/348025#discussion_r1826913885)

And later _in the same file_ `doc/build-helpers/testers.chapter.md`:

> [In fact it would have been better to state the bottom line of the comparison here to save people time when reading.](https://github.com/NixOS/nixpkgs/pull/348025#discussion_r1826913985)

Now I don't know if you are [complaining about the duplication](https://github.com/NixOS/nixpkgs/pull/348025#discussion_r1827407783) _in the same file_ or if you are [agreeing with this duplication](https://github.com/NixOS/nixpkgs/pull/348025#discussion_r1827480362) by stating

> . . . but I think it should be mentioned in the same file twice, in less then 10 lines distance between each mention.

IMHO I have no problem duplicating this.

(BTW the duplication around multiple files is not relevant here.)

B

doronbehar replied:
> > . . . but I think it should be mentioned in the same file twice, in less then 10 lines distance between each mention.
> 
> IMHO I have no problem duplicating this.

Oh I'm so sorry @AndersonTorres . I forgot to negate that sentence, I think it _shouldn't_ be mentioned in the same file twice, in a such a close distance of <10 lines apart.

AndersonTorres replied:
Done.

doronbehar commented on doc/hooks/versionCheckHook.section.md:
```diff
@@ -16,20 +22,67 @@ stdenv.mkDerivation (finalAttrs: {
+| Execution environment                  | The hook uses `env --ignore-environment --chdir=/` to run the executables in an environment as clean as possible, but there is no way to change its behavior. | The tester executes in an identical environment of a consumer, independent from the building environment.                                                                                            ...
+| Rebuild after modification             | The hook rebuilds the package when the hook is modified, since it is a phase running during the build time of the package.                                    | The tester does not require rebuilding the package, since it is a derivation.                                                                                                                        ...
+| Overhead during package building       | Negligible. Although running during the build time of the package, the hook is lean and executes few commands.                                                | Zero, since the tester is a derivation that runs after package building.                                                                                                                             ...
+| Content-addressed-derivation awareness | Since it runs during the building, the hook does not deal with failures happening after building, like rewritings that happen post installation.              | Since it runs after the building, the tester detects failures at this time.                                                                                                                          ...
```
I must say that I don't understand this line of comparison at all. But maybe it is just me, who is not very well familiar with content addressed derivations...

AndersonTorres replied:
The citation of CA will change to something more general.

doronbehar commented on doc/hooks/versionCheckHook.section.md:
```diff
@@ -16,20 +22,67 @@ stdenv.mkDerivation (finalAttrs: {
+| Execution by OfBorg CI tool            | Yes, since the hook is a phase running during the build time of the package.                                                                                  | OfBorg does run the tester for _directly affected packages only_; transitive dependencies are ignored by default, requiring extra commands like `@ofborg build dependency1.tests dependency2.tests . ...
+| Execution by `nixpkgs-review` tool     | Yes, since the hook is a phase running during the build time of the package.                                                                                  | No, since the tool has no support for executing `passthru.tests`. [^1]                                                                                                                               ...
+| Package breakage awareness             | Loud and clear as soon as the hook is reached, since it is a phase running during the build time of the package.                                              | Requires specific command to be noticed, e.g. `nix-build -A pkg.tests.version`, since it is a derivation dependent on the package                                                                    ...
+| Transitive package breakage            | Never ever goes unnoticed by `nixpkgs-review`, since the tool executes the transitive dependencies.                                                           | Since human beings are prone to forget the duty of running passthru tests and `nixpkgs-review` has no support for running it, it certainly goes unnoticed.                                           ...
```
Haha this:

> human beings are prone to forget the duty of running passthru tests

Is a very blaming phrasing.. I don't think there is a consensus that people should run `passthru.tests` of all packages they use on their system regularly or whatever. It is not exactly clear in which context it would be correct to blame someone about such a breakage. My opinion is that no one should be blamed for not doing something that our tooling doesn't allow them to easily do.

AndersonTorres replied:
I will remove references to less-than-omniscient things in the next iteration.

doronbehar commented on doc/hooks/versionCheckHook.section.md:
```diff
@@ -16,20 +22,67 @@ stdenv.mkDerivation (finalAttrs: {
+
+
+| Item                                   | `versionCheckHook`                                                                                                                                            | `testers.testVersion`                                                                                                                                                                                ...
+|:---------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------:|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ...
```
Nit: Looks better IMO when lines are aligned to the left.

AndersonTorres replied:
I will remove these lines, at least until I discover why the manual was not built.

doronbehar commented on doc/hooks/versionCheckHook.section.md:
```diff
@@ -16,20 +22,67 @@ stdenv.mkDerivation (finalAttrs: {
+| Content-addressed-derivation awareness | Since it runs during the building, the hook does not deal with failures happening after building, like rewritings that happen post installation.              | Since it runs after the building, the tester detects failures at this time.                                                                                                                          ...
+| Execution by OfBorg CI tool            | Yes, since the hook is a phase running during the build time of the package.                                                                                  | OfBorg does run the tester for _directly affected packages only_; transitive dependencies are ignored by default, requiring extra commands like `@ofborg build dependency1.tests dependency2.tests . ...
+| Execution by `nixpkgs-review` tool     | Yes, since the hook is a phase running during the build time of the package.                                                                                  | No, since the tool has no support for executing `passthru.tests`. [^1]                                                                                                                               ...
+| Package breakage awareness             | Loud and clear as soon as the hook is reached, since it is a phase running during the build time of the package.                                              | Requires specific command to be noticed, e.g. `nix-build -A pkg.tests.version`, since it is a derivation dependent on the package                                                                    ...
+| Transitive package breakage            | Never ever goes unnoticed by `nixpkgs-review`, since the tool executes the transitive dependencies.                                                           | Since human beings are prone to forget the duty of running passthru tests and `nixpkgs-review` has no support for running it, it certainly goes unnoticed.                                           ...
```
In general, it is not very clear from reading the 1st column of these 2 lines, what exactly is being compared. I understand how you wanted to make it more concise then in the table I wrote in my previous review, but it is not exactly clear what does "breakage" mean, in which context? In Nixpkgs PRs? Or in a user's NixOS system? Or in Hydra?

I admit that even in my table the last line could have been a bit clearer, (I hope you will forgive me for demanding better clarity when this is something that everyone will read v.s something only you read when I suggest something for your PR).

AndersonTorres replied:
> but it is not exactly clear what does "breakage" mean

Breakage is any event that triggers the respective tool.

> , in which context? In Nixpkgs PRs? Or in a user's NixOS system? Or in Hydra?

It does not matter who is the one that triggers the event.

`versionCheckHook` is triggered by anyone that tries to build the package containing it, whether it is a `nixos-rebuild  switch` executed by a GitHub Action or a user running commands in a Powershell instance.

The same can be said about `passthru.tests.version` _mutatis mutandis_.

doronbehar replied:
> > but it is not exactly clear what does "breakage" mean
> 
> Breakage is any event that triggers the respective tool.

The respective tool is nixpkgs-review or ofborg? Or the package itself?

> > , in which context? In Nixpkgs PRs? Or in a user's NixOS system? Or in Hydra?
> 
> It does not matter who is the one that triggers the event.

I disagree, because in any of the above contexts there are different chances of a package with `versionCheckHook` or a package's `passthru.tests.version` derivation to be built. This distinction is a very important point for comparison, and the first column of these 2 lines don't make it clear to which context you are referring to.
 
> `versionCheckHook` is triggered by anyone that tries to build the package containing it, whether it is a `nixos-rebuild switch` executed by a GitHub Action or a user running commands in a Powershell instance.

Thanks for adding to the list of possible contexts :). To summarize, the possible contexts I see are:

- A user builds locally packages in a modified Nixpkgs checkout.
- `nixpkgs-review` in a PR.
- Someone running `@ofborg build` in a PR.
- A user running `nixos-rebuild`.
- A GitHub action running `nixos-rebuild`, and possibly other checks if configured to do so.

doronbehar commented on doc/stdenv/passthru.chapter.md:
```diff
@@ -75,17 +75,16 @@ The Nixpkgs systems for continuous integration [Hydra](https://hydra.nixos.org/)
-- If the `versionCheckPhase` (the phase defined by [`versionCheckHook`](#versioncheckhook)) fails, it triggers a failure which can't be ignored if you use the package, or if you find out about it in a [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review) report.
-- Sometimes packages become silently broken - meaning they fail to launch but their build passes because they don't perform any tests in the `checkPhase`. If you use this tool infrequently, such a silent breakage may rot in your system / profile configuration, and you will not notice the failure until you will want to use this package. Testing such basic functionality ensures you have to deal wit ...
-- When you open a PR, [ofborg](https://github.com/NixOS/ofborg)'s CI _will_ run `passthru.tests` of [packages that are directly changed by your PR (according to your commits' messages)](https://github.com/NixOS/ofborg?tab=readme-ov-file#automatic-building), but if you'd want to use the [`@ofborg build`](https://github.com/NixOS/ofborg?tab=readme-ov-file#build) command for dependent packages, you  ...
+Note: Especifically for version-testing, Nixpkgs provides [`testers.testVersion`](#tester-testVersion), traditionally used as `passthru.tests.version` attribute, as well as the `versionCheckHook` mentioned above.
```
A link to the comparison is required here, along with the bottom line which agreeably states that "given the current limitations of Nixpkgs's tooling and CI demonstrated above, `versionCheckHook` fits better in most of typical situations."

AndersonTorres replied:
Done.

doronbehar reviewed (commented):
Getting better!

doronbehar commented on doc/hooks/versionCheckHook.section.md:
```diff
@@ -16,20 +22,65 @@ stdenv.mkDerivation (finalAttrs: {
+| Customization                             | Besides the attributes described above, `versionCheckHook` provides no other methods for controlling it.                                                                | `tests.version` has access to the whole feature set of Nixpkgs, like multiple version test derivations and tools outside the package inputs. |
+| Execution time                            | `versionCheckHook` is executed automatically whenever the package is executed, e.g. `nix-build -A pkg`, since it is a phase executed during the package build time.     | `tests.version` is executed only when explicitly called, e.g. `nix-build -A pkg.tests.version`; it is executed after package build time.     |
+| Execution environment                     | `versionCheckHook` uses `env --ignore-environment --chdir=/` to run the executables in an environment as clean as possible, but there is no way to change its behavior. | `tests.version` executes in an environment identical to that of a consumer, independent from the package build environment.                  |
+| Rebuild after modification                | Modifying `versionCheckHook` triggers the package rebuild.                                                                                                              | `tests.version` does not rebuild the package, since it runs after package build time.                                                        |
```
Another accuracy nit:

```suggestion
| Rebuild after modification of hook/test derivation | Modifying `versionCheckHook` triggers rebuilds for all packages using it.                                                                                                              | Doesn't require rebuilding the package, as the test derivation is independent of it. |
```

doronbehar commented on doc/hooks/versionCheckHook.section.md:
```diff
@@ -16,20 +22,65 @@ stdenv.mkDerivation (finalAttrs: {
+| Execution time                            | `versionCheckHook` is executed automatically whenever the package is executed, e.g. `nix-build -A pkg`, since it is a phase executed during the package build time.     | `tests.version` is executed only when explicitly called, e.g. `nix-build -A pkg.tests.version`; it is executed after package build time.     |
+| Execution environment                     | `versionCheckHook` uses `env --ignore-environment --chdir=/` to run the executables in an environment as clean as possible, but there is no way to change its behavior. | `tests.version` executes in an environment identical to that of a consumer, independent from the package build environment.                  |
+| Rebuild after modification                | Modifying `versionCheckHook` triggers the package rebuild.                                                                                                              | `tests.version` does not rebuild the package, since it runs after package build time.                                                        |
+| Overhead during package build time        | Negligible. Although executed during the the package build time, `versionCheckHook` is lean and executes few commands.                                                  | Zero, since `tests.version` is a derivation that runs after package build time.                                                              |
```
Can be more accurate:

```suggestion
| Overhead during package build time        | Negligible - executes the main package's command 1-2 times, during the package's build time. | Zero, since `tests.version` is an independent derivation. |
```

It's the 2nd time you write "a derivation that runs after a package build time" or something roughly the same - I hope you are OK with the fact I emphasize that not necessarily people will build (and not run) that derivation, because people might forget to do it by themselves or look at ofborg's attempts (which are [sometimes](https://github.com/NixOS/nixpkgs/pull/353607#issuecomment-2456283386) inaccurate).

doronbehar commented on doc/hooks/versionCheckHook.section.md:
```diff
@@ -16,20 +22,65 @@ stdenv.mkDerivation (finalAttrs: {
+| Execution environment                     | `versionCheckHook` uses `env --ignore-environment --chdir=/` to run the executables in an environment as clean as possible, but there is no way to change its behavior. | `tests.version` executes in an environment identical to that of a consumer, independent from the package build environment.                  |
+| Rebuild after modification                | Modifying `versionCheckHook` triggers the package rebuild.                                                                                                              | `tests.version` does not rebuild the package, since it runs after package build time.                                                        |
+| Overhead during package build time        | Negligible. Although executed during the the package build time, `versionCheckHook` is lean and executes few commands.                                                  | Zero, since `tests.version` is a derivation that runs after package build time.                                                              |
+| Failure detection time                    | `versionCheckHook` runs during the package build time, therefore it does not deal with failures happening after that.                                                   | `tests.version` detects failures after package build time, like rewritings and relocations.                                                  |
```
I don't understand this line of comparison. Do you mean "when will `versionCheckHook`/`testVersion` tell us a package is failing? Is this supposed to be the content-addressed derivations related line? What do you mean by failures happening after builds? Again perhaps my lack of CA derivations knowledge is holding me back from understanding.

doronbehar commented on doc/hooks/versionCheckHook.section.md:
```diff
@@ -16,20 +22,65 @@ stdenv.mkDerivation (finalAttrs: {
+| Rebuild after modification                | Modifying `versionCheckHook` triggers the package rebuild.                                                                                                              | `tests.version` does not rebuild the package, since it runs after package build time.                                                        |
+| Overhead during package build time        | Negligible. Although executed during the the package build time, `versionCheckHook` is lean and executes few commands.                                                  | Zero, since `tests.version` is a derivation that runs after package build time.                                                              |
+| Failure detection time                    | `versionCheckHook` runs during the package build time, therefore it does not deal with failures happening after that.                                                   | `tests.version` detects failures after package build time, like rewritings and relocations.                                                  |
+| Package breakage awareness                | Loud and clear as soon as `versionCheckHook` is reached during the package build time.                                                                                  | Requires specific commands to be noticed, e.g. `nix-build -A pkg.tests.version`.                                                             |
```
Small nit now:

```suggestion
| Package breakage awareness                | Loud and clear as soon as `versionCheckHook` is reached during the package build time.                                                                                  | Requires running specific commands to be noticed, e.g. `nix-build -A pkg.tests.version`.                                                             |
```

doronbehar commented on doc/hooks/versionCheckHook.section.md:
```diff
@@ -16,20 +22,65 @@ stdenv.mkDerivation (finalAttrs: {
+| Overhead during package build time        | Negligible. Although executed during the the package build time, `versionCheckHook` is lean and executes few commands.                                                  | Zero, since `tests.version` is a derivation that runs after package build time.                                                              |
+| Failure detection time                    | `versionCheckHook` runs during the package build time, therefore it does not deal with failures happening after that.                                                   | `tests.version` detects failures after package build time, like rewritings and relocations.                                                  |
+| Package breakage awareness                | Loud and clear as soon as `versionCheckHook` is reached during the package build time.                                                                                  | Requires specific commands to be noticed, e.g. `nix-build -A pkg.tests.version`.                                                             |
+| OfBorg CI                                 | `versionCheckHook` will be executed by OfBorg, since ofBorg automatically executes the packages affected by the pull request (PR).                                      | OfBorg supports automatic execution of `passthru.tests` for _packages directly affected_ by the PR.                                          |
+| OfBorg CI, transitive dependencies        | `versionCheckHook` will executed by OfBorg, since it builds transitive dependencies of a package affected by the PR. (????)                                             | OfBorg does not support automatic execution of `passthru.tests` for transitive dependencies,  requiring extra commands for this.             |
```
Hmm maybe each line should explain better to which ofborg behavior are we referring to. Here's my suggestion:

```suggestion
| OfBorg's [automatic building](https://github.com/NixOS/ofborg?tab=readme-ov-file#automatic-building) treatment | Will be executed as it is part of the package's build phase | Will be executed too as part of the _directly_ changed packages' `passthru.tests` |
| [`@ofborg build`](https://github.com/NixOS/ofborg?tab=readme-ov-file#commands) treatment | Will be executed too as it is part of the package's build phase | Will be executed only if you'd remember to specify the test derivation with e.g `@ofborg build pkg.passthru.tests.version` |
```

doronbehar commented on doc/hooks/versionCheckHook.section.md:
```diff
@@ -16,20 +22,65 @@ stdenv.mkDerivation (finalAttrs: {
+| Package breakage awareness                | Loud and clear as soon as `versionCheckHook` is reached during the package build time.                                                                                  | Requires specific commands to be noticed, e.g. `nix-build -A pkg.tests.version`.                                                             |
+| OfBorg CI                                 | `versionCheckHook` will be executed by OfBorg, since ofBorg automatically executes the packages affected by the pull request (PR).                                      | OfBorg supports automatic execution of `passthru.tests` for _packages directly affected_ by the PR.                                          |
+| OfBorg CI, transitive dependencies        | `versionCheckHook` will executed by OfBorg, since it builds transitive dependencies of a package affected by the PR. (????)                                             | OfBorg does not support automatic execution of `passthru.tests` for transitive dependencies,  requiring extra commands for this.             |
+| `nixpkgs-review`                          | `versionCheckHook` will be executed by `nixpkgs-review`.                                                                                                                | `nixpkgs-review` has no support for executing `passthru.tests` at all. [^1]                                                                  |
+| `nixpkgs-review`, transitive dependencies | `versionCheckHook` will be executed by `nixpkgs-review`, since the tool reaches transitive dependencies automatically.                                                  | `nixpkgs-review` has no support for executing `passthru.tests` at all,  even for transitive dependencies. [^1]                               |
```
Here it is weird to split these to 2 lines, as nixpkgs-review always acts upon all affected packages.

doronbehar reviewed (commented):
While going through I noticed that this is still sort of WIP? Anyway here are my comments. The general structure looks great, only the comparison table could use some rephrasing.

doronbehar commented:
> 2. Before squashing the `[wip]` commits, I let them this way to ease the review.

I stopped reading this commit by commit long ago TBH :), but I'll take that into consideration next time.

> 1. Table comparison is cute when formatted but horrible to be read in source code

True! I used https://tableconvert.com/ to post the initial table.

> 3\. The error reported by CI says nothing at all.

I agree. Seems like a bug. Let's ask for help from the @NixOS/documentation-team .

AndersonTorres commented on doc/hooks/versionCheckHook.section.md:
````diff
@@ -16,20 +22,65 @@ stdenv.mkDerivation (finalAttrs: {
 ```
 
-Note that for [`buildPythonPackage`](#buildpythonpackage-function) and [`buildPythonApplication`](#buildpythonapplication-function), `doInstallCheck` is enabled by default.
+:::
````
The triple dot?

It closes the `::: {.example . . .}` block.
I don't remember if this is mandatory or merely recommended, but example code is usually enclosed in blocks like these.

doronbehar commented:
> > 3. The error reported by CI says nothing at all.
> 
> I agree. Seems like a bug. Let's ask for help from the @NixOS/documentation-team .

Any member of @NixOS/documentation-team can help with the CI error we see here?

fricklerhandwerk commented:
Hm, could be something related to the custom extensions. I recommend bisecting by removing the footnote and table, and check if it goes away. Maybe it's the footnote in the table.

AndersonTorres committed:
[wip] multiple tables

doronbehar commented:
@AndersonTorres see also https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/69 - it might be related.
